gateway/src/apicast/policy/caching/policy.lua

```diff
@@ -0,0 +1,94 @@
+--- Caching policy
+-- Configures a cache for the authentication calls against the 3scale backend.
+-- The 3scale backend can authorize (status code = 200) and deny (status code =
+-- 4xx) calls. When it fails, it returns a 5xx code.
+-- This policy support three kinds of caching:
+--   - Strict: it only caches authorized calls. Denied and failed calls
+--     invalidate the cache entry.
+--   - Resilient: caches authorized and denied calls. Failed calls do not
+--     invalidate the cache. This allows us to authorize and deny calls
+--     according to the result of the last request made even when backend is
+--     down.
+--   - None: disables caching.
+
+local policy = require('apicast.policy')
+local _M = policy.new('Caching policy')
+
+local new = _M.new
+
+local function strict_handler(cache, cached_key, status_code, ttl)
+  if status_code == 200 then
+    ngx.log(ngx.INFO, 'apicast cache write key: ', cached_key,
+                      ', ttl: ', ttl, ' sub: ')
+    cache:set(cached_key, 200, ttl or 0)
+  else
+    ngx.log(ngx.NOTICE, 'apicast cache delete key: ', cached_key,
+                        ' cause status ', status_code)
+    cache:delete(cached_key)
+  end
+end
+
+local function resilient_handler(cache, cached_key, status_code, ttl)
+  if status_code and status_code < 500 then
+    ngx.log(ngx.INFO, 'apicast cache write key: ', cached_key,
+                      ' status: ', status_code, ', ttl: ', ttl)
+
+    cache:set(cached_key, status_code, ttl or 0)
+  end
+end
+
+local function disabled_cache_handler()
+  return function() end
```

Comment:
This is going to allocate function every time this is executed. It is not necessary no?

Reply:
Right. We can choose one in the initializer.

Reply:
This line can be removed, right? 
Or maybe we could add a debug log line saying that it skipped cache.

Reply:
:+1: